### ISSUE TITLE
[Feature] - Add ValidateTypeAttribute

### ISSUE THREAD

**Issue by vexx32** [PowerShell/PowerShell#8288]
Currently there are several core cmdlets and no shortage of custom commands in public usage which declare parameters of type `[object]` (or `[psobject]`) and then after parameter binding decide whether or not input is usable, with varying success (as is practically always the case where multiple custom implementations may be present).

We are able to specify, currently, only a single possible parameter type. We _also_ have, for some convenience, the `[PSTypeName()]` attribute. However, given that PSObjects can be given completely arbitrary type names, this is at best a matter of convenience and ease of use where PSCustomObject usage is concerned. It cannot and does not attempt to validate that objects have the runtime type they may 'pretend' to have with an inserted PSTypeName.

This leads to unclear help documentation and parameter displays, and the user must either read the expanded help for such parameters specifically or the cmdlet as a whole, because the parameter diagrams are simply inaccurate. 

As such, we have a bit of a gap here that we can fill. We could make use of a parameter attribute (suggested name: `ValidateTypeAttribute` aka `[ValidateType()]`) which would accept an _array_ of `Type` objects and/or _strings_ (provided those strings map to a resolvable runtime type; could be declared by the type's fullname or shortname/accelerator name if one is available).

This would provide necessary programmatic metadata for improvements to `Get-Help` whereby such parameters would in the syntax diagram explicitly be flagged as accepting (for example) _either_ `[string]`, _or_ `[hashtable]` or whatever else it might expect.

I am willing to hash out the code for this, if it turns out to be something y'all like the concept of. 😄

**Comment by iSazonov** [comment]
Look interesting but we need discuss samples where it would be useful. Specially a bind error is terminating but current behavior can be to write non-terminating error.

**Comment by bergmeister** [comment]
I like the idea. It would be nice if the attribute could also do type inference, i.e. being able to specify a type (either interface or implementation) and if the type is implementing that interface or deriving from that implementation then it accepts it as well.

**Comment by romero126** [comment]
There are several instances I can think of where PowerShell users would find this useful. As well as a tool for internal development.

This would be great passing a [PSObject] to a specific overload in C# a great toolkit for future work in module builders.

[ValidateType([int], [int32], [string], [float])]
PSObject foo;

where you could take 
var x = foo.BaseObject;
calloverload(x);

This would be really good for that.

Also could also allow for type casting capabilities for desired type's without a need for a switch case or try catch all your error handling would be done at a validation level.

I think this would be invaluable.

**Comment by vexx32** [comment]
I don't think this will allow us to bypass compile-time constraints in that way; as I understand it, method overloads are determined at *compile* time, so it won't be possible to dynamically select which overload you're using... however, it might pave the way for a generic set of overloads to be created for a purpose like this. I'm not clear how this would function, in all honesty, I'm hazy about the possible use cases on that end... but if it's a possibility it'd definitely prove interesting to look at!

**Comment by iSazonov** [comment]
By the way, we already have attributes to describe the source and return types. Although they are not used.

**Comment by KirkMunro** [comment]
I worry this may promote bad design.

For commands that are out there that take an object or a `PSObject`, should those parameters be that generic?

In core PowerShell commands, many of those should be generic because they work against objects. A few other ones (e.g. `Get-Acl` and `Set-Acl` from WindowsPowerShell) accept a `PSObject` and then have some other criteria that they require (in the case of `Get-Acl` or `Set-Acl`, the presence of a `GetSecurityDescriptor` or `SetSecurityDescriptor` method is required).

What might help here are specific examples demonstrating why PowerShell would benefit from a specific attribute for this, and why ValidateScript is not sufficient for those scenarios, or why a different design of the command syntax wouldn't be a better option.

**Comment by vexx32** [comment]
Look at something like `Select-Object`'s `-Property` parameter:

```
-Property <Object[]>
    Specifies the properties to select.  These properties are added as `NoteProperty` members to the output objects.
    Wildcards are permitted.

    The value of the Property parameter can be a new calculated property. To create a calculated, property, use a hash
    table. Valid keys are:

    - Name (or Label) <string>

    - Expression <string> or <script block>

    Required?                    false
    Position?                    0
    Default value                None
    Accept pipeline input?       False
    Accept wildcard characters?  true
```
Note the ambiguous nature of the parameter type, there is a brief mention that you can use property names, which to those of us already knowing how it works, we know that's a string, and then a mention of the hashtable syntax that can be used.

The parameters here are, as far as I know, evaluated at runtime to validate them. We could shortcut this quite a bit if we simply had a `[ValidateType([string], [hashtable])]`, and the help could look significantly cleaner, specifying clearly that only one of those two types are acceptable. The additional help text then need only specify the acceptable keys for hashtable input.

**Comment by KirkMunro** [comment]
Thanks, that's a great example, just what I was looking for, and I get what you're after now.

Food for thought: I think this feature would be better implemented using an "Allow" attribute instead of a "Validate" attribute. e.g. Instead of `[ValidateType([string], [hashtable])]`, I think it would be more self documenting if the syntax was written like this: `[AllowType([string],[hashtable])]`, and if the parser only allowed the attribute to be used for parameters of type `[object]` or `[psobject]`.

**Comment by vexx32** [comment]
`Allow` attributes are not restrictive but permissive, so I don't think it's semantically appropriate to break that convention, given part of this suggestion is to help with validating the input types and remove the need to do so at runtime.
